Fix keyward arguments warning in `MiddlewareStack#build`

kamipo opened the pull request:
This is an alternative of 6b633a2823feb09f51daac57f410ee559a56195a.

`ActionDispatch::Static` middleware requires to be passed `**kwargs`.

https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/lib/rails/application/default_middleware_stack.rb#L27
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/actionpack/lib/action_dispatch/middleware/static.rb#L112-L116

But `ActionDispatch::Session::CookieStore` middleware requires to be
passed actual `session_options` to `options.merge!(cookie_only: true)`
destructively.

(Ideally passed args should not be mutated, but for now the middleware
intentionally murate the args and testing that.)

https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/lib/rails/application/default_middleware_stack.rb#L65
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb#L59-L61
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/test/application/configuration_test.rb#L1557-L1563
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/test/application/middleware/session_test.rb#L337-L341

To address both case, we need to evaluate args for building app in the
`use` method context (that method args in the context have last args
whether kwargs or hash).

kamipo committed:
Fix keyward arguments warning in `MiddlewareStack#build`

This is an alternative of 6b633a2823feb09f51daac57f410ee559a56195a.

`ActionDispatch::Static` middleware requires to be passed `**kwargs`.

https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/lib/rails/application/default_middleware_stack.rb#L27
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/actionpack/lib/action_dispatch/middleware/static.rb#L112-L116

But `ActionDispatch::Session::CookieStore` middleware requires to be
passed actual `session_options` to `options.merge!(cookie_only: true)`
destructively.

(Ideally passed args should not be mutated, but for now the middleware
intentionally murate the args and testing that.)

https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/lib/rails/application/default_middleware_stack.rb#L65
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb#L59-L61
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/test/application/configuration_test.rb#L1557-L1563
https://github.com/rails/rails/blob/0b3c710d35199d187dbfddaefbb9ab2752ea6498/railties/test/application/middleware/session_test.rb#L337-L341

To address both case, we need to evaluate args for building app in the
`use` method context (that method args in the context have last args
whether kwargs or hash).

eregon commented:
I'm a bit confused by this solution.
It seems `ActionDispatch::Session::CookieStore` is always passed positional arguments (the last one being a Hash), never keyword arguments, so `ruby2_keywords` should solve that easily.

First, the PR seems to fix `use` but not `unshift`, `insert`, `insert_after` and `swap`.

In https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html I showed multiple ways to fix ActionDispatch::MiddlewareStack and added some tests (https://github.com/eregon/rails/commit/f414dd52eb47046beb998acb1bdb4282293ec1b3).

Here is how it looks like with `ruby2_keywords`: https://github.com/eregon/rails/commit/44a80647f5
Since `ruby2_keywords` has the behavior to flag any kwargs passed to a `*args` parameter, it works for those methods, and delegates kwargs as kwargs in `build` which uses `klass.new(app, *args, &block)` (also needs to be `*args` there but it already is).

What do you think of that? Should I make a PR with that commit? I think it's a simpler and more complete solution.

cc @jeremyevans @mame

---

Interesting for me is that your solution is somehow similar to my `pass_keywords` solution: https://github.com/eregon/rails/commit/8b0625ed68
Instead of flagging the kwargs it instead works locally inside the method and also works inside blocks.
For that to work, we need that `build_middleware` takes `*args` and not `args`, but that's a private method so it seems fine.

kamipo commented:
> First, the PR seems to fix `use` but not `unshift`, `insert`, `insert_after` and `swap`.

Yes, most 2.7 warnings in building app are ignored in #37215, I've fixed `unshift`, `insert`, `insert_after` and `swap` and reverted #37215 in #38091.

rafaelfranca commented:
Is there any way to fix this issue without calling `ruby2_keywords` or any of those ugly methods? I prefer conditionally defining methods depending on the Ruby version than calling those methods.

eregon commented:
@kamipo Thanks, but why are you passing a block? It shouldn't be needed, `ruby2_keywords` on those 5 methods should be all that's needed. Can you show a test that fails without that block passed to `build_middleware`?

eregon commented:
@rafaelfranca I think using `...` inside `eval` would the only way then, as shown in https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html#the--operator
Because delegation is not done directly here, we'd need an approach similar to https://github.com/eregon/rails/commit/8b0625ed68, replacing
`pass_keywords def mymethod(klass, *args, &block)` with `def mymethod(...)`.
However that's not good enough because `insert` accepts an `index` argument, and `...` doesn't accept leading arguments in 2.7: https://bugs.ruby-lang.org/issues/16378
There are workaround like `first(...)` to extract the first argument out of it but it's not pretty either.

https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/ is clear that `ruby2_keywords` is the only recommended way (I don't like either, but that's how it is).

That blog post also mentions another way under `A compatible delegation that works on Ruby 2.6, 2.7 and Ruby 3`.
```ruby
if RUBY_VERSION < "2.7"
  def delegate(*args, &block)
    target(*args, &block)
  end
else
  def delegate(*args, **kwargs, &block)
    target(*args, **kwargs, &block)
  end
end
```
might work good enough here but there at least one case broken with it (`delegate({})` passes no arguments at all to `target`).

kamipo commented:
@eregon Thanks for the pointing out. I see the passing a block is unneeded, I've removed that in 51908b641f078342f3adcf3242990c2d73e291f5.

eregon commented:
@kamipo Thanks, it looks good now :)

From your commit message:
> I misunderstood the effect of `ruby2_keywords`, the effect is not
> local scope, the effect still alive even where you delegate.

Indeed, I argued that a lexical `pass_keywords` (detailed [here](https://eregon.me/blog/2019/11/10/the-delegation-challenge-of-ruby27.html#pass_keywords)) would be easier to understand than some hidden state like `ruby2_keywords`. OTOH, `ruby2_keywords` is more concise for this case.
